File: app/api/canvas_api.py
"""
Canvas API Client for EaselyBot
Handles Canvas LMS API integration including token validation and assignment fetching
"""

import logging
import requests
from typing import Dict, List, Optional, Any
from config.settings import CANVAS_BASE_URL, CANVAS_API_VERSION
from app.utils.retry_helper import exponential_backoff_retry, is_retryable_http_error

logger = logging.getLogger(__name__)


class CanvasAPIClient:
    """Canvas LMS API client for token validation and data fetching"""
    
    def __init__(self):
        self.base_url = CANVAS_BASE_URL
        self.api_version = CANVAS_API_VERSION
        
    @exponential_backoff_retry(
        retries=3,
        initial_delay=1.0,
        max_delay=10.0,
        exceptions=(requests.exceptions.RequestException, requests.exceptions.Timeout)
    )
    def _make_request(self, token: str, endpoint: str, method: str = 'GET', params: Dict = None) -> Optional[Any]:
        """
        Make authenticated request to Canvas API with retry logic
        
        Args:
            token: Canvas access token
            endpoint: API endpoint (without /api/v1 prefix)
            method: HTTP method
            params: Query parameters
            
        Returns:
            Response data or None if failed
        """
        try:
            url = f"{self.base_url}/api/{self.api_version}/{endpoint}"
            headers = {
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json'
            }
            
            response = requests.request(
                method=method,
                url=url,
                headers=headers,
                params=params or {},
                timeout=10
            )
            
            if response.status_code == 200:
                return response.json()
            elif is_retryable_http_error(response.status_code):
                # Raise exception to trigger retry for retryable errors
                raise requests.exceptions.RequestException(
                    f"Retryable Canvas API error: {response.status_code} - {response.text}"
                )
            else:
                logger.error(f"Canvas API error (non-retryable): {response.status_code} - {response.text}")
                return None
                
        except requests.exceptions.RequestException as e:
            # Let the decorator handle retries for these exceptions
            logger.warning(f"Canvas API request failed, will retry: {str(e)}")
            raise
    
    def validate_token(self, token: str) -> Dict[str, Any]:
        """
        Validate Canvas access token by fetching user profile
        
        Args:
            token: Canvas access token to validate
            
        Returns:
            Dict with validation result:
            {
                'valid': bool,
                'user_info': dict or None,
                'error_message': str or None
            }
        """
        try:
            # Test token by getting user profile
            user_data = self._make_request(token, 'users/self')
            
            if user_data:
                return {
                    'valid': True,
                    'user_info': {
                        'id': user_data.get('id'),
                        'name': user_data.get('name'),
                        'email': user_data.get('email'),
                        'login_id': user_data.get('login_id')
                    },
                    'error_message': None
                }
            else:
                return {
                    'valid': False,
                    'user_info': None,
                    'error_message': 'Invalid token or Canvas server error'
                }
                
        except Exception as e:
            logger.error(f"Token validation error: {str(e)}")
            return {
                'valid': False,
                'user_info': None,
                'error_message': f'Validation error: {str(e)}'
            }
    
    def get_user_courses(self, token: str) -> List[Dict[str, Any]]:
        """
        Get user's active courses
        
        Args:
            token: Canvas access token
            
        Returns:
            List of course dictionaries
        """
        try:
            courses = self._make_request(token, 'courses', params={
                'enrollment_state': 'active',
                'per_page': 100
            })
            
            if courses:
                return [{
                    'id': course.get('id'),
                    'name': course.get('name'),
                    'course_code': course.get('course_code'),
                    'term': course.get('term', {}).get('name', 'Unknown')
                } for course in courses]
            
            return []
            
        except Exception as e:
            logger.error(f"Error fetching courses: {str(e)}")
            return []
    
    def get_assignments(self, token: str, limit: int = 500) -> List[Dict[str, Any]]:
        """
        Get upcoming assignments across all courses
        
        Args:
            token: Canvas access token
            limit: Maximum number of assignments to return
            
        Returns:
            List of assignment dictionaries
        """
        try:
            logger.info(f"Fetching assignments from Canvas API (limit: {limit})")
            
            # First get user's courses
            courses = self.get_user_courses(token)
            logger.info(f"Found {len(courses)} courses for user")
            
            if not courses:
                logger.warning("No courses found for user")
                return []
            
            assignments = []
            
            for course in courses:  # Get ALL courses, not just first 10
                logger.debug(f"Fetching assignments for course: {course['name']} ({course['id']})")
                
                course_assignments = self._make_request(
                    token, 
                    f'courses/{course["id"]}/assignments',
                    params={
                        'per_page': 100,  # Get more assignments per course
                        'order_by': 'due_at'
                    }
                )
                
                if course_assignments:
                    logger.info(f"Found {len(course_assignments)} assignments in {course['name']}")
                    
                    for assignment in course_assignments:
                        # Only include assignments with due dates
                        if assignment.get('due_at'):
                            assignments.append({
                                'id': assignment.get('id'),
                                'title': assignment.get('name'),
                                'course_name': course['name'],
                                'course_code': course['course_code'],
                                'due_date': assignment.get('due_at'),
                                'description': assignment.get('description', ''),
                                'points_possible': assignment.get('points_possible'),
                                'submission_types': assignment.get('submission_types', []),
                                'html_url': assignment.get('html_url')
                            })
                        else:
                            logger.debug(f"Skipping assignment '{assignment.get('name')}' - no due date")
                else:
                    logger.info(f"No assignments found in {course['name']}")
            
            logger.info(f"Total assignments with due dates found: {len(assignments)}")
            
            # Sort by due date and return ALL assignments (no limit)
            assignments.sort(key=lambda x: x['due_date'] or '')
            
            logger.info(f"Returning ALL {len(assignments)} assignments (no limit)")
            return assignments
            
        except Exception as e:
            logger.error(f"Error fetching assignments: {str(e)}")
            return []
    
    def get_upcoming_assignments(self, token: str, days_ahead: int = 14) -> List[Dict[str, Any]]:
        """
        Get assignments due in the next X days (FUTURE ONLY, not past)
        
        Args:
            token: Canvas access token
            days_ahead: Number of days to look ahead
            
        Returns:
            List of upcoming assignment dictionaries
        """
        try:
            from datetime import datetime, timedelta, timezone
            now = datetime.now(timezone.utc)
            cutoff_date = now + timedelta(days=days_ahead)
            
            # Get ALL assignments from ALL courses
            all_assignments = self.get_assignments(token, limit=500)
            upcoming = []
            
            for assignment in all_assignments:
                if assignment['due_date']:
                    try:
                        # Parse Canvas datetime format
                        due_date = datetime.fromisoformat(assignment['due_date'].replace('Z', '+00:00'))
                        # ONLY include FUTURE assignments (due_date > now) and within cutoff
                        if now <= due_date <= cutoff_date:
                            upcoming.append(assignment)
                    except (ValueError, AttributeError):
                        continue
            
            return upcoming  # Return ALL upcoming, no limit
            
        except Exception as e:
            logger.error(f"Error fetching upcoming assignments: {str(e)}")
            return []
    
    def create_calendar_event(self, token: str, title: str, start_at: str, end_at: str = None, description: str = None) -> Dict[str, Any]:
        """
        Create a calendar event in Canvas (user's personal calendar)
        
        Args:
            token: Canvas access token
            title: Event title
            start_at: Event start datetime in ISO format
            end_at: Event end datetime (optional, defaults to start_at)
            description: Event description (optional)
            
        Returns:
            Created event data or error dict
        """
        try:
            # If no end time specified, make it same as start (all-day or point event)
            if not end_at:
                end_at = start_at
            
            # Prepare event data
            event_data = {
                'calendar_event': {
                    'context_code': 'user_self',  # Personal calendar
                    'title': title,
                    'start_at': start_at,
                    'end_at': end_at
                }
            }
            
            if description:
                event_data['calendar_event']['description'] = description
            
            # Make POST request to create event
            url = f"{self.base_url}/api/{self.api_version}/calendar_events"
            headers = {
                'Authorization': f'Bearer {token}',
                'Content-Type': 'application/json'
            }
            
            response = requests.post(
                url=url,
                headers=headers,
                json=event_data,
                timeout=10
            )
            
            if response.status_code in [200, 201]:
                event = response.json()
                logger.info(f"Created Canvas calendar event: {title}")
                return {
                    'success': True,
                    'event_id': event.get('id'),
                    'title': event.get('title'),
                    'start_at': event.get('start_at'),
                    'html_url': event.get('html_url')
                }
            else:
                logger.error(f"Failed to create calendar event: {response.status_code} - {response.text}")
                return {
                    'success': False,
                    'error': f"Canvas API error: {response.status_code}"
                }
                
        except Exception as e:
            logger.error(f"Error creating calendar event: {str(e)}")
            return {
                'success': False,
                'error': str(e)
            }


# Global Canvas API client instance
canvas_client = CanvasAPIClient()


def validate_canvas_token(token: str) -> Dict[str, Any]:
    """
    Convenience function to validate a Canvas token
    
    Args:
        token: Canvas access token
        
    Returns:
        Validation result dictionary
    """
    return canvas_client.validate_token(token)


def fetch_user_assignments(token: str, limit: int = 500) -> List[Dict[str, Any]]:
    """
    Convenience function to fetch user assignments
    
    Args:
        token: Canvas access token
        limit: Maximum assignments to return (default 200 for all)
        
    Returns:
        List of assignment dictionaries
    """
    # Get assignments for the next 90 days to ensure we get all relevant ones
    assignments = canvas_client.get_upcoming_assignments(token, days_ahead=90)
    # Return all assignments, not truncated
    return assignments

File: app/api/canvasApi.js
/**
 * Canvas LMS API Integration
 * Fetches assignments and course data using user-specific Canvas tokens
 */

const axios = require('axios');
const { CANVAS_BASE_URL, CANVAS_API_VERSION } = require('../../config/settings');
const { getUser } = require('../database/supabaseClient');
const moment = require('moment-timezone');

/**
 * Create Canvas API client for a specific user
 * @param {string} canvasToken - User's Canvas API token
 * @param {string} canvasUrl - Canvas instance URL (optional, uses default if not provided)
 * @returns {Object} Axios instance configured for Canvas API
 */
function createCanvasClient(canvasToken, canvasUrl = null) {
    const baseURL = `${canvasUrl || CANVAS_BASE_URL}/api/${CANVAS_API_VERSION}`;
    
    return axios.create({
        baseURL,
        headers: {
            'Authorization': `Bearer ${canvasToken}`,
            'Accept': 'application/json+canvas-string-ids',
            'Content-Type': 'application/json'
        },
        timeout: 30000 // 30 second timeout
    });
}

/**
 * Get user's Canvas token from database
 * @param {string} facebookId - Facebook user ID
 * @returns {Promise<Object|null>} Object with token and URL or null if not found
 */
async function getUserCanvasCredentials(facebookId) {
    try {
        console.log(`\n🔑 Fetching Canvas credentials for user ${facebookId}`);
        const user = await getUser(facebookId);
        
        if (!user) {
            console.log(`❌ User ${facebookId} not found in database`);
            return null;
        }
        
        if (!user.canvas_token) {
            console.log(`❌ No Canvas token found for user ${facebookId}`);
            return null;
        }
        
        // Log token details (safely)
        console.log(`✅ Found Canvas token for user ${facebookId}`);
        console.log(`   Token length: ${user.canvas_token.length} characters`);
        console.log(`   Token preview: ${user.canvas_token.substring(0, 10)}...`);
        console.log(`   Canvas URL: ${user.canvas_url || CANVAS_BASE_URL}`);
        console.log(`   Canvas User ID: ${user.canvas_user_id || 'Not stored'}`);
        
        return {
            token: user.canvas_token,
            url: user.canvas_url || CANVAS_BASE_URL,
            userId: user.canvas_user_id
        };
    } catch (error) {
        console.error(`❌ Error fetching Canvas credentials for user ${facebookId}:`, error.message);
        return null;
    }
}

/**
 * Fetch all active courses for a user
 * @param {string} facebookId - Facebook user ID
 * @returns {Promise<Array>} Array of course objects
 */
async function fetchUserCourses(facebookId) {
    const credentials = await getUserCanvasCredentials(facebookId);
    
    if (!credentials) {
        throw new Error('Canvas not connected. Please provide your Canvas API token first.');
    }
    
    const client = createCanvasClient(credentials.token, credentials.url);
    
    try {
        const response = await client.get('/courses', {
            params: {
                enrollment_state: 'active',
                per_page: 100
            }
        });
        
        return response.data || [];
    } catch (error) {
        console.error('Error fetching Canvas courses:', error.response?.data || error.message);
        throw new Error('Failed to fetch courses from Canvas. Please check your Canvas token.');
    }
}

/**
 * Fetch upcoming assignments for a user
 * @param {string} facebookId - Facebook user ID
 * @param {number} daysAhead - Number of days to look ahead (default: 7)
 * @returns {Promise<Array>} Array of assignment objects
 */
async function fetchUpcomingAssignments(facebookId, daysAhead = 7) {
    console.log(`\n📚 Fetching upcoming assignments for user ${facebookId} (next ${daysAhead} days)`);
    
    const credentials = await getUserCanvasCredentials(facebookId);
    
    if (!credentials) {
        console.log(`❌ Cannot fetch assignments - no Canvas token found`);
        throw new Error('Canvas not connected. Please provide your Canvas API token first.');
    }
    
    console.log(`🔗 Creating Canvas API client with user's token`);
    const client = createCanvasClient(credentials.token, credentials.url);
    const assignments = [];
    const endDate = moment().add(daysAhead, 'days').toISOString();
    
    try {
        // First, get all active courses
        console.log(`📡 Calling Canvas API: GET /courses (using token: ${credentials.token.substring(0, 10)}...)`);
        const coursesResponse = await client.get('/courses', {
            params: {
                enrollment_state: 'active',
                per_page: 100
            }
        });
        
        const courses = coursesResponse.data || [];
        console.log(`📁 Found ${courses.length} active courses`);
        
        // For each course, fetch assignments
        for (const course of courses) {
            try {
                console.log(`  📡 Fetching assignments for course: ${course.name}`);
                const assignmentsResponse = await client.get(`/courses/${course.id}/assignments`, {
                    params: {
                        per_page: 100,
                        order_by: 'due_at',
                        bucket: 'upcoming' // Get upcoming assignments
                    }
                });
                
                const courseAssignments = assignmentsResponse.data || [];
                console.log(`    Found ${courseAssignments.length} assignments in ${course.name}`);
                
                // Filter assignments by due date and add course info
                const upcomingAssignments = courseAssignments
                    .filter(assignment => {
                        if (!assignment.due_at) return false;
                        const dueDate = moment(assignment.due_at);
                        return dueDate.isAfter(moment()) && dueDate.isBefore(endDate);
                    })
                    .map(assignment => ({
                        ...assignment,
                        course_name: course.name,
                        course_code: course.course_code,
                        course_id: course.id
                    }));
                
                assignments.push(...upcomingAssignments);
            } catch (courseError) {
                console.error(`Error fetching assignments for course ${course.id}:`, courseError.message);
                // Continue with other courses even if one fails
            }
        }
        
        // Sort by due date
        assignments.sort((a, b) => moment(a.due_at).diff(moment(b.due_at)));
        
        console.log(`✅ Successfully fetched ${assignments.length} total assignments using user's token`);
        return assignments;
    } catch (error) {
        console.error('❌ Error fetching Canvas assignments:', error.response?.data || error.message);
        if (error.response?.status === 401) {
            throw new Error('Canvas token is invalid or expired. Please update your Canvas token.');
        }
        throw new Error('Failed to fetch assignments from Canvas. Please check your Canvas token.');
    }
}

/**
 * Fetch assignments for this week
 * @param {string} facebookId - Facebook user ID
 * @returns {Promise<Array>} Array of assignment objects
 */
async function fetchThisWeekAssignments(facebookId) {
    console.log(`\n📅 Fetching this week's assignments for user ${facebookId}`);
    
    const credentials = await getUserCanvasCredentials(facebookId);
    
    if (!credentials) {
        console.log(`❌ Cannot fetch assignments - no Canvas token found`);
        throw new Error('Canvas not connected. Please provide your Canvas API token first.');
    }
    
    console.log(`🔗 Using Canvas token from database for user ${facebookId}`);
    const client = createCanvasClient(credentials.token, credentials.url);
    const assignments = [];
    
    // Calculate week boundaries
    const startOfWeek = moment().startOf('week');
    const endOfWeek = moment().endOf('week');
    console.log(`📆 Week range: ${startOfWeek.format('MMM D')} - ${endOfWeek.format('MMM D')}`);
    
    try {
        // Get all active courses
        const coursesResponse = await client.get('/courses', {
            params: {
                enrollment_state: 'active',
                per_page: 100
            }
        });
        
        const courses = coursesResponse.data || [];
        
        // Fetch assignments for each course
        for (const course of courses) {
            try {
                const assignmentsResponse = await client.get(`/courses/${course.id}/assignments`, {
                    params: {
                        per_page: 100,
                        order_by: 'due_at'
                    }
                });
                
                const courseAssignments = assignmentsResponse.data || [];
                
                // Filter for this week's assignments
                const weekAssignments = courseAssignments
                    .filter(assignment => {
                        if (!assignment.due_at) return false;
                        const dueDate = moment(assignment.due_at);
                        return dueDate.isBetween(startOfWeek, endOfWeek, 'day', '[]');
                    })
                    .map(assignment => ({
                        ...assignment,
                        course_name: course.name,
                        course_code: course.course_code,
                        course_id: course.id
                    }));
                
                assignments.push(...weekAssignments);
            } catch (courseError) {
                console.error(`Error fetching assignments for course ${course.id}:`, courseError.message);
            }
        }
        
        // Sort by due date
        assignments.sort((a, b) => moment(a.due_at).diff(moment(b.due_at)));
        
        return assignments;
    } catch (error) {
        console.error('Error fetching this week\'s assignments:', error.response?.data || error.message);
        throw new Error('Failed to fetch this week\'s assignments from Canvas.');
    }
}

/**
 * Fetch a single assignment by ID
 * @param {string} facebookId - Facebook user ID
 * @param {string} courseId - Canvas course ID
 * @param {string} assignmentId - Canvas assignment ID
 * @returns {Promise<Object>} Assignment object
 */
async function fetchAssignment(facebookId, courseId, assignmentId) {
    const credentials = await getUserCanvasCredentials(facebookId);
    
    if (!credentials) {
        throw new Error('Canvas not connected. Please provide your Canvas API token first.');
    }
    
    const client = createCanvasClient(credentials.token, credentials.url);
    
    try {
        const response = await client.get(`/courses/${courseId}/assignments/${assignmentId}`);
        return response.data;
    } catch (error) {
        console.error('Error fetching assignment:', error.response?.data || error.message);
        throw new Error('Failed to fetch assignment from Canvas.');
    }
}

/**
 * Test Canvas API connection with user's token
 * @param {string} canvasToken - Canvas API token
 * @param {string} canvasUrl - Canvas instance URL (optional)
 * @returns {Promise<Object>} User profile if successful
 */
async function testCanvasConnection(canvasToken, canvasUrl = null) {
    const client = createCanvasClient(canvasToken, canvasUrl);
    
    try {
        const response = await client.get('/users/self');
        return {
            success: true,
            user: response.data
        };
    } catch (error) {
        console.error('Canvas connection test failed:', error.response?.data || error.message);
        return {
            success: false,
            error: 'Invalid Canvas token or URL. Please check your credentials.'
        };
    }
}

/**
 * Format assignment for display
 * @param {Object} assignment - Canvas assignment object
 * @returns {string} Formatted assignment string
 */
function formatAssignment(assignment) {
    const dueDate = moment(assignment.due_at);
    const now = moment();
    const daysUntilDue = dueDate.diff(now, 'days');
    const hoursUntilDue = dueDate.diff(now, 'hours');
    
    let timeString;
    if (daysUntilDue > 1) {
        timeString = `in ${daysUntilDue} days`;
    } else if (daysUntilDue === 1) {
        timeString = 'tomorrow';
    } else if (hoursUntilDue > 0) {
        timeString = `in ${hoursUntilDue} hours`;
    } else {
        timeString = 'OVERDUE';
    }
    
    return `📚 ${assignment.name}\n` +
           `📖 ${assignment.course_name}\n` +
           `⏰ Due: ${dueDate.format('MMM D, h:mm A')} (${timeString})\n` +
           `${assignment.html_url ? `🔗 ${assignment.html_url}` : ''}`;
}

/**
 * Format multiple assignments for display
 * @param {Array} assignments - Array of Canvas assignment objects
 * @returns {string} Formatted assignments string
 */
function formatAssignmentsList(assignments) {
    if (!assignments || assignments.length === 0) {
        return '✨ No assignments due! You\'re all caught up!';
    }
    
    const formattedList = assignments
        .slice(0, 10) // Limit to 10 assignments to avoid message being too long
        .map((assignment, index) => {
            const dueDate = moment(assignment.due_at);
            const shortDate = dueDate.format('MMM D');
            const shortTime = dueDate.format('h:mm A');
            return `${index + 1}. ${assignment.name}\n   📖 ${assignment.course_code || assignment.course_name}\n   ⏰ ${shortDate} at ${shortTime}`;
        })
        .join('\n\n');
    
    const header = assignments.length === 1 
        ? '📚 You have 1 assignment:\n\n' 
        : `📚 You have ${assignments.length} assignments:\n\n`;
    
    return header + formattedList;
}

module.exports = {
    createCanvasClient,
    getUserCanvasCredentials,
    fetchUserCourses,
    fetchUpcomingAssignments,
    fetchThisWeekAssignments,
    fetchAssignment,
    testCanvasConnection,
    formatAssignment,
    formatAssignmentsList
};
